src/crypto/slow-hash.c

```diff
@@ -1767,3 +1841,127 @@ void cn_slow_hash(const void *data, size_t length, char *hash, int variant, int
 }
 
 #endif
+
+#define SEEDHASH_EPOCH_BLOCKS	2048
+#define SEEDHASH_EPOCH_LAG	64
+
+bool rx_needhash(const uint64_t height, uint64_t *seedheight) {
+  uint64_t s_height =  (height <= SEEDHASH_EPOCH_BLOCKS+SEEDHASH_EPOCH_LAG) ? 0 :
+                       (height - SEEDHASH_EPOCH_LAG - 1) & ~(SEEDHASH_EPOCH_BLOCKS-1);
+  rx_state *rx_sp;
+  rx_s_toggle = (s_height & SEEDHASH_EPOCH_BLOCKS) != 0;
+  *seedheight = s_height;
+  rx_sp = &rx_s[rx_s_toggle];
+  if (!s_height && !rx_sp->rs_cache)
+    return true;
+  return (rx_sp->rs_height != s_height);
+}
+
+typedef struct seedinfo {
+  unsigned long si_start;
+  unsigned long si_count;
+} seedinfo;
+
+static THREAD_RTYPE rx_seedthread(void *arg) {
+  seedinfo *si = arg;
+  rx_state *rx_sp = &rx_s[rx_s_toggle];
+  randomx_init_dataset(rx_dataset, rx_sp->rs_cache, si->si_start, si->si_count);
+  return NULL;
+}
+
+void rx_seedhash(const uint64_t height, const char *hash, const int miners) {
+  randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+  rx_state *rx_sp = &rx_s[rx_s_toggle];
+  MUTEX_LOCK(rx_mutex);
+  if (rx_sp->rs_height == height && rx_sp->rs_cache != NULL)
+    goto leave;
+  if (use_v4_jit())
+    flags |= RANDOMX_FLAG_JIT;
+  if (rx_sp->rs_cache == NULL) {
+    rx_sp->rs_cache = randomx_alloc_cache(flags | RANDOMX_FLAG_LARGE_PAGES);
+    if (rx_sp->rs_cache == NULL)
+      rx_sp->rs_cache = randomx_alloc_cache(flags);
+  }
+  if (rx_sp->rs_cache != NULL) {
+    randomx_init_cache(rx_sp->rs_cache, hash, 32);
+    rx_sp->rs_height = height;
+    if (miners) {
+      if (rx_dataset == NULL) {
+        rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_LARGE_PAGES);
+        if (rx_dataset == NULL)
+          rx_dataset = randomx_alloc_dataset(RANDOMX_FLAG_DEFAULT);
+      }
+      if (rx_dataset != NULL) {
+        if (miners > 1) {
+          unsigned long delta = randomx_dataset_item_count() / miners;
+          unsigned long start = 0;
+          int i;
+          seedinfo *si;
+          THREAD_TYPE *st;
+          si = malloc(miners * sizeof(seedinfo));
+          if (si == NULL)
+            goto leave;
+          st = malloc(miners * sizeof(THREAD_TYPE));
+          if (st == NULL) {
+            free(si);
+            goto leave;
+          }
+          for (i=0; i<miners; i++) {
+            si[i].si_start = start;
+            si[i].si_count = delta;
+            start += delta;
+          }
+          si[i-1].si_count = randomx_dataset_item_count() - si[i-1].si_start;
+          for (i=1; i<miners; i++) {
+            THREAD_CREATE(st[i], rx_seedthread, &si[i]);
+          }
+          randomx_init_dataset(rx_dataset, rx_sp->rs_cache, 0, si[0].si_count);
+          for (i=1; i<miners; i++) {
+            THREAD_JOIN(st[i]);
+          }
+          free(st);
+          free(si);
+        } else {
+          randomx_init_dataset(rx_dataset, rx_sp->rs_cache, 0, randomx_dataset_item_count());
+        }
+        flags |= RANDOMX_FLAG_FULL_MEM;
```

Comment:
The problem here is that only the first VM will be created with this flag. All the other VMs will be created in light mode. Also monerod seems to release the VM that finds a block, so the "FULL_MEM" VM will be destroyed quickly and only the light memory VMs will mine. This is the reason why the hashrate spikes at the beginning, but then drops to a much lower value after you find the first block.

Reply:
OK, your explanation of the behavior makes sense, but I don't see why this would be. All of the mining threads are going to call rx_seedhash() with miners > 0. The mining threads are started and stopped on demand, so there's no leftover thread-local state for them to be picking up.

Reply:
Should be all good now